File: 02 November 2022/C++/Enemy.cpp
//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++

class Solution
{
    public:
        int largestArea(int n,int m,int k,vector<vector<int>> &enemy)
        {
            //  code here
            vector<int> v1, v2;
            for(auto &it: enemy){
                v1.push_back(it[0]);
                v2.push_back(it[1]);
            }
            sort(v1.begin(), v1.end());
            sort(v2.begin(), v2.end());
            v1.push_back(n+1);
            v2.push_back(m+1);
            int rowMax = INT_MIN;
            int start = 0;
            for(auto &it: v1){
                rowMax = max(rowMax, it-start-1);
                start = it;
            }
            int colMax = INT_MIN;
            start = 0;
            for(auto &it: v2){
                colMax = max(colMax, it-start-1);
                start = it;
            }
            return rowMax * colMax;
        }
};

//{ Driver Code Starts.

int main()
{
    int t;
    cin>>t;
    while(t--)
    {
        int n,m;
        cin>>n>>m;
        int k;
        cin>>k;
        vector<vector<int>> e(k,vector<int>(2));
        for(int i=0;i<k;i++)
            cin>>e[i][0]>>e[i][1];
        Solution a;
        cout<<a.largestArea(n,m,k,e)<<endl;
    }
    return 0;
}
// } Driver Code Ends

File: 02 November 2022/Python/Enemy.py
from typing import List

class Solution:
    def largestArea(self,n:int,m:int,k:int, enemy : List[List[int]]) -> int:
        # code here
        rows=[]
        cols=[]
        if k==0:
            return n*m
        for i in e:
            rows.append(i[0])
            cols.append(i[1])
        rows.sort()
        cols.sort()
        #print(rows,cols)
        #This is for calculating the area above the "first blocked" row and the area
        # below the last blocked row
        le=max(rows[0]-1,n-rows[k-1])
        #This is for calculating the area left to the "first blocked" cols adn the area
        #right to the "last blocked" cols
        wid=max(cols[0]-1,m-cols[k-1])
        #If there are blocked rows in between the first and last blocked rows
        #And If there are blocked cols in between the left and right cols
        for i in range(1,k):
            le=max(le,(rows[i]-(rows[i-1])-1))
        for i in range(1,k):
            wid=max(wid,(cols[i]-(cols[i-1])-1))
        return le*wid
        

#{ 
 # Driver Code Starts
class IntArray:
    def __init__(self) -> None:
        pass
    def Input(self,n):
        arr=[int(i) for i in input().strip().split()]#array input
        return arr
    def Print(self,arr):
        for i in arr:
            print(i,end=" ")
        print()



class IntMatrix:
    def __init__(self) -> None:
        pass
    def Input(self,n,m):
        matrix=[]
        #matrix input
        for _ in range(n):
            matrix.append([int(i) for i in input().strip().split()])
        return matrix
    def Print(self,arr):
        for i in arr:
            for j in i:
                print(j,end=" ")
            print()


if __name__=="__main__":
    t = int(input())
    for _ in range(t):
        
        a=IntArray().Input(3)
        n,m,k=a[0],a[1],a[2]
        
        e=IntMatrix().Input(a[2], 2)
        
        obj = Solution()
        res = obj.largestArea(n,m,k, e)
        
        print(res)
        

# } Driver Code Ends